ini.c

```diff
@@ -46,22 +46,34 @@ static int ini_property(struct ini *ini, char *key, char *value)
 	if (!ini->prop_cb)
 		return 0;
 
-	// If the value begins with 'xresource:` then query Xresources for the actual value.
-    if (strncmp(value, "xresource:", strlen("xresource:")) == 0) {
-        if (!ini->database) {
-            fatal("Unable to access Xresources.");
-            return -EINVAL;
-        }
-
-        char *resource;
-        if (xcb_xrm_resource_get_string(ini->database, value + strlen("xresource:"), NULL, &resource) == 0) {
-            value = strdup(resource);
-            free(resource);
-        } else {
-            fatal("invalid Xresource key: \"%s\"", value);
-            return -EINVAL;
-        }
-    }
+	// If the value begins with 'xresource:` then query Xresources for the actual value or load an optional default.
+	if (strncmp(value, "xresource:", strlen("xresource:")) == 0) {
+		if (!ini->database) {
+			fatal("Unable to access Xresources.");
+			return -EINVAL;
+		}
+
+		// Add null terminator so that xcb_xrm_resource_get_string knows where to stop.
+		int buffer_size = strlen(value);
+		char *default_value = strchr(value, ' ');
+		if (default_value) {
+			if (*(default_value + 1) != '\0') {
+				default_value++;
+				default_value[-1] = '\0';
+			}
+		}
+
+		char *resource;
+		if (xcb_xrm_resource_get_string(ini->database, value + strlen("xresource:"), NULL, &resource) == 0) {
+			strncpy(value, resource, buffer_size);
```

Comment:
This is the only part I had some concern about. I'm reusing the `value` buffer, since that is allocated on the stack and will be cleaned up automatically.

However, if the contents of the resource lookup are longer than the resource ID (+ default value), then it could be truncated when copied back into the `value` object. I think I see a solution to this problem and will update the PR.

Looking at the existing code, I'm almost certain that the `strdup` call is a memory leak.

Reply:
Thanks for the details and the fix @winder!